### ISSUE TITLE
GPU access

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#4675]
## Problem

Containers run by Dagger can't access their machine's GPUs. This makes using Dagger impractical for some pipelines that would otherwise be a great fit: video encoding, ML inference, training, etc.

## Solution

Add GPU access to the Dagger API.

## Design considerations

The bottleneck is probably buildkit, which [doesn't seem to support GPU access](https://github.com/NVIDIA/nvidia-container-toolkit/issues/122). But I haven't looked into it beyond a quick web search.

**Comment by shykes** [comment]
This is a requirement for @jpetazzo :)

**Comment by shykes** [comment]
Thoughts @sipsma @vito @aluzzardi ? Is this a small / medium / large?

**Comment by jpetazzo** [comment]
_..ooOO( Thinking out loud )_

The first thing that I'll try is to refactor my encoding/inference containers so that they can work on stdio, which should let me run them as "host containers" through a bind-mounted Docker socket, which should be (fairly?) easy in dagger (if I'm not mistaken).

The second thing would be to perhaps see if I can somehow... inject the nvidia container runtime in the buildkit container and convince buildkit to use it? Hmm. That might be tricky 😅

(Also, just to give you a bit of context: currently, my workflow revolves around a big "data" directory which is shared by all the stages of the pipeline - speech recognition, OCR, barcode detection, video transcoding... Each "job" is approximately 5-10 GB of input data + 5-10 GB of output data + 5-10 GB of intermediary files. In a perfect world, this would be all stored in an object store and each stage would get data / process / put data; and once the pipeline is prod-ready, this is probably what will happen; but for the time being, working on local files saves a ton of time - especially since getting/pushing 5 GB from my local machine each time I want to do a round of testing isn't great 😅)

**Comment by kjuulh** [comment]
I guess some stuff like running machine learning models, transcoding, rendering etc could be a use case. 

You could setup a pipeline to take in assets run them through blender, or do some training with it.

That said, this is definitely a niche use-case =D

---

I've previously used either the nvidia container runtime (now deprecated) and now the built in support in linux (in docker itself, or probably containerd).

**Comment by sipsma** [comment]
Yeah my understanding of existing GPU support for containers is that it mostly amounts to on-the-fly modifications of the oci runtime config for the container (mount in certain devices, config files and drivers that are host specific, etc.).

Dagger already has its own shim inserted between buildkit and runc for that exact same purpose of modifying the oci runtime config on the fly, so that would be the perfect place to start with adding gpu hooks. We can then look into upstreaming this to buildkit.

The more annoying part of this is that the dagger engine is itself a docker container by default, isolated from the host. It should already have access to the device files of the host since it runs w/ `--privileged`, but it will also need access to the rest of the possible files. The choice might be between:
1. Unconditionally always running the engine w/ flags that give it access to anything it might need (nvidia, amd, intel, etc.)
2. Making this a configuration knob for spinning up the engine that the user has to be aware of and set
3. Something like what @kjuulh suggested where the gpu is served over a socket rather than mounted into the container (IIUC, I didn't even know that was possible, quite cool if so!)

My immediate instinct is that `1.` is preferable but I don't know yet how feasible it is, a little bit more research needed. Either way this should be possible.

**Comment by robertu94** [comment]
Use Case: My team develops GPU based data compression libraries (among other things) while building them can be done with stubs, testing them requires actual GPUs.

**Comment by robertu94** [comment]
For what it's worth there is a standard API for adding GPU devices across vendors already supported with containerd, docker, and podman https://github.com/container-orchestrated-devices/container-device-interface.  I guess you just need provide a method to pass a the equivalent of a `--device` flag.

**Comment by mircubed** [comment]
related to #4684

**Comment by mircubed** [comment]
mentioned the need for GPU access in this blog post - https://robertu94.github.io/2023/04/24/refactoring-ci/cd-for-a-moderately-large-c-code-base.html cc @gerhard

**Comment by gerhard** [comment]
What are your thoughts on this @lukemarsden?

**Comment by samalba** [comment]
> 1. Unconditionally always running the engine w/ flags that give it access to anything it might need (nvidia, amd, intel, etc.)

From what I understand from @sipsma's comment here, the dagger scheme could auto add the right flags when gpu(s) are detected on the host. And given that we run in privileged mode, it seems acceptable. That way, we would not need an extra config to pass to the pipeline anytime it requires gpu access, it would work transparently.

**Comment by mircubed** [comment]
For anyone following this thread, @lukemarsden will demo using Dagger with an MLOps use case at the next community call (June 29th). He is working on a workaround for GPU support that you can use until this issue is completed. Register below before you forget ;) 

Register here: https://dagger-io.zoom.us/webinar/register/3916868482633/WN_USQjVBGXT0SWhNMvqYVvCA

**Comment by samalba** [comment]
Assigning this @matiasinsaurralde who agreed to give a first shot. I'll summarize a mini-plan (note that this is based on personal opinion, and recent conversation with a user, not authoritative).

### Step 1 - Add basic GPU access (like docker run)

The idea is to do provide something similar than docker: `docker run --gpu gpu_request`. The gpu_request is either `all` or a device id like `--gpu device=0`.

In the dagger CLI, we could replicate this with (python):

```python
# client -> dagger.Client
container_gpu = client.container()._from("some-image").with_gpu("all")
# Or select one specific device
container_gpu = client.container()._from("some-image").with_gpu(device=0)
```

This approach is as good as what docker supports, which is not great. GPUs are not multi-tenant. As a result most users end up mapping one workflow to 1 physical device. You end up paying for expensive gpus that are under-utilized.

However it's a good first step.

### Step 2 - Add a memory consumption target

This is an addition to Step 1, not a replacement. The idea is to extend the API with every step, not break it.

```python
# client -> dagger.Client
container_gpu = client.container()._from("some-image").with_gpu(memory_max="1gb")
```

This call will go through all available GPUs, look for available `memory_max` and select the gpu device that satisfies the memory constraint. If the constraint can't be met (not enough memory, or no device available), it raises an error.

This approach is an improvement over step 1 because there is now a primitive that allows to use a single device for multiple workflow (based on memory available).

However it's not perfect because the memory lookup happens at the container creation, and the available memory can quickly change over time.

### Step 3 - Monitor the device memory usage for a given time before selecting a device

```python
# client -> dagger.Client
container_gpu = client.container()._from("some-image").with_gpu(memory_max="1gb", threshold="5", timeout="30")
```

This call will go through all available GPUs, look for available `memory_max`, if the device is found, it will monitor the memory at a short interval (TBD) and ensure the constraint remains fullfilled for `threshold` (seconds) before returning the device. If the device being monitored does not meet the contraint below threshold (available memory < memory_max), then the GPU selection restarts until we reach the timeout.

If we reach the timeout and no device had enough available memory for the given threshold, an error is raised.


This is open to feedback. In the meantime I would start by implementing Step 1.

cc @matiasinsaurralde @aluzzardi @shykes @sipsma

**Comment by shykes** [comment]
I'm excited to see this moving :)

One note on the plan: our "step 1" must abstract away the synchronization of device access. In other words, it must take care of allocating the next available device, *or blocking until a device is available*. This is different from docker, which passes the synchronization work to the caller.

In this version of step 1, you still map one job to one device; but at least you don't have to write brittle and host-specific synchronization wrapper scripts.

**Comment by robertu94** [comment]
A few items to think about:

To the best of my knowledge neither Nvidia, Amd, nor Intel GPUs currently provide an api to actually enforce limits on the memory usage on the GPU per process. I keep holding out hope that Linux might mandate something that at least Intel and Amd would need to support, but I don't see Nvidia implementing anything like that soon. I like the idea of making scheduling decisions based on Memory and would like the feature, but I suspect it will be hard to implement enforceably and consistently. At best you might be able to programmatically access the Nvidia smi (or equivalent output) and kill a dagger created pid if we happen to see exceeds its memory allowance. I'm not sure if there are proper public APIs even to query the GPU memory used by a particular  process outside of nvidia-smi (but I haven't looked hard). 

Second, it is possible to virtually partition GPUs à la SRIOV, but generally only by root (possibly only on boot, but I don't recall) and only on data center class cards such as Nvidia's a100 or v100. Like other GPU partioning APIs you can't infinitely slice and dice things and need to partition according to hardware subdivisions. 

Third, which device may actually be important. On some systems the dual network interface are directly connected to 2 of say 6 GPUs (let's call them 0 and 3). The user may want to require that certain jobs be placed on those GPUs to do things like GPU direct (or the AMD:Intel equivalent) efficiently. Users still make programatic decisions about how many of which vendors GPU are near which resources using libraries like hwloc, but they'll make that decision on their own about how to map their jobs to them. IMHO, the simplest and most useful v1 API is "all", devices=[0,1]. Anything smarter is likely to be really wrong for some users.

**Comment by shykes** [comment]
> To the best of my knowledge neither Nvidia, Amd, nor Intel GPUs currently provide an api to actually enforce limits on the memory usage on the GPU per process. I keep holding out hope that Linux might mandate something that at least Intel and Amd would need to support, but I don't see Nvidia implementing anything like that soon. I like the idea of making scheduling decisions based on Memory and would like the feature, but I suspect it will be hard to implement enforceably and consistently. At best you might be able to programmatically access the Nvidia smi (or equivalent output) and kill a dagger created pid if we happen to see exceeds its memory allowance. I'm not sure if there are proper public APIs even to query the GPU memory used by a particular process outside of nvidia-smi (but I haven't looked hard).

> > Second, it is possible to virtually partition GPUs à la SRIOV, but generally only by root (possibly only on boot, but I don't recall) and only on data center class cards such as Nvidia's a100 or v100. Like other GPU partioning APIs you can't infinitely slice and dice things and need to partition according to hardware subdivisions.

This particular feature is based on an idea described by @jqueguiner, which may be best described as a "production-grade hack". The idea is to *not* rely on virtual partitioning, but implement a crude, best-effort memory partitioning system instead by relying on standard OS primitives, and knowledge of how memory mapping works in practice. I won't claim to understand this design in full detail, but my understanding is that it's reliable enough to yield considerable real-world performance gains; but it does require solid handling of errors and retries, since there's nothing stopping a job from using more memory than was "fake allocated" by the crude scheduler. What makes this model operationally doable (again, my understanding) is that over-use results in a clean crash, which allows for clean error handling, retries etc.

TLDR: this memory allocation feature is firmly in the "experimental / possibly maybe" column.

> Third, which device may actually be important. On some systems the dual network interface are directly connected to 2 of say 6 GPUs (let's call them 0 and 3). The user may want to require that certain jobs be placed on those GPUs to do things like GPU direct (or the AMD:Intel equivalent) efficiently. Users still make programatic decisions about how many of which vendors GPU are near which resources using libraries like hwloc, but they'll make that decision on their own about how to map their jobs to them. IMHO, the simplest and most useful v1 API is "all", devices=[0,1]. Anything smarter is likely to be really wrong for some users. What might make more sense is provide callbacks for users to decide which GPUs (or to block and wait for more of them), based on what's currently available from Dsggers perspective.

Do the device numbers have a portable meaning (in the way that, say, like file descriptors 0, 1 and 2 have a portable meaning)? If so, we can do this without a problem.

However, if those numbers *don't* have a portable meaning (my guess), then we can't expose them in the Dagger API, even as a stopgap, because that would be a one-way door that would result in breaking changes later etc. However, we could still implement a portable device enumeration API, with all the attributes needed for programmatic decision exposed, in a portable way.

**Comment by robertu94** [comment]
As I understand, the numbers are consistent from the perspective of a single vendor from run to run (maybe not boot to boot), but not necessarily portable between machines or vendor (other than the "first" device is 0 for some vendor specific definition of first). This is why I mentioned using the portable and vendor agnostic hwloc library to determine the actual topology is, and choose the corresponding ids to place things as desired. hwloc provides the mostly vendor agnostic abstractions to know Nvidia GPU 0 is of vendor Nvidia has 4GB of Ram, and is directly connected to socket 1, and AMD GPU 0 has 8 GB of ram on socket 2, and SYCL calls the AMD GPU 0 and the Nvidia one GPU 1 (made up example). I say mostly here because it does allow getting vendor specific values, but there are generic things too. 

Where this becomes more complicated (as I hinted) is if you have 2+ different vendors or programing frameworks in the same machine (this is uncommon in production). Each vendor calls one of their own cards GPU 0. This is why OpenCL/SYCL/OpenMp etc have a two level namespace for GPUs of the form vendor:deviceid rather than using the PCI address (which *may* be unique, but I don't think is guaranteed for example in disaggregated architectures) because some libraries can utilize GPUs from multiple vendors meaning card0 may have different ids depending on what library you use. Docker gets away with using just integers  because very few people have more than one vendor's discrete GPUs in their system and you can just pick the integers to line up with how the vendor enumerates them.

**Comment by matiasinsaurralde** [comment]
Following up with the first idea for basic GPU access. I've spent some time exploring the implementation details around Docker and trying to modify Dagger to follow a similar approach:

- In Docker initialization happens around `nvidia_linux` which it's invoked as soon as the daemon starts. Attempts to resolve the path for `nvidia-container-runtime-hook` (typically found in `/usr/bin`).
- If the hook binary is found on the current path: (i) a capabilities set [is initialized](https://github.com/moby/moby/blob/master/pkg/capabilities/caps.go) and (ii) `registerDeviceDriver` is called taking the capabitilies set and a function called `setNvidiaGPUs`.
- `setNvidiaGPUs` initializes environment variables (`NVIDIA_VISIBLE_DEVICES`, `NVIDIA_DRIVER_CAPABILITIES`) and attaches a [`prestart` hook](https://github.com/opencontainers/runtime-spec/blob/main/config.md#posix-platform-hooks) to the runtime spec, pointing to the `nvidia-container-runtime-hook` binary.
**Note:** Even though `prestart` hooks look deprecated seems is still very much in use, even in Docker's implementation.
- `setNvidiaGPUs` is actually called when [`WithDevices`](https://github.com/moby/moby/blob/3fae7ea16f9b80e58119d59394b8b5da0887d533/daemon/oci_linux.go#L911) is used, [`HostDevices`/`WithHostDevices`](https://github.com/containerd/containerd/blob/96de54db4385279a5f8474b5de6658f6336b6fd8/oci/utils_unix.go#L43) is used to populate the host devices on the runtime spec. A `container.DeviceRequest` is constructed and passed to the previously registered `setNvidiaGPUs` hook.
- Spec for a container that's using `gpus=all` in Docker contains the following block (full spec in case there's some other relevant info [here](https://gist.github.com/matiasinsaurralde/9199552e1bbec961b711c8bb33862dc6)):
```json
            "DeviceRequests": [
                {
                    "Driver": "",
                    "Count": -1,
                    "DeviceIDs": null,
                    "Capabilities": [
                        [
                            "gpu"
                        ]
                    ],
                    "Options": {}
                }
            ],
```
- `gpus=0`:
```json
            "DeviceRequests": [
                {
                    "Driver": "",
                    "Count": 0,
                    "DeviceIDs": null,
                    "Capabilities": [
                        [
                            "gpu"
                        ]
                    ],
                    "Options": {}
                }
            ],
```
- To summarize: initial detection of the presence of the NV container hook is performed on the host when daemon starts. Later a hook is registered to modify the runtime spec, this initializes environment variables and ends up calling the NV container hook. Still unsure on the exact step where `WithDevices` is called to trigger this logic.

On the Dagger side I did some hacking around the following ideas:
- Hacked the engine so that `nvidia-container-runtime-hook` is available from the `setupBundle` -where the runtime spec is built-.
- Modified it to attempt to call the hook as:
```diff
        spec.Process.Env = keepEnv
 
+       hook := specs.Hook{
+               //Path: "/usr/bin/nvidia-container-runtime-hook",
+               Path: "/bin/sh",
+               // Path: dest,
+               Args: []string{
+                       // "ls",
+                       // "-lah",
+                       "/bin/sh",
+                       "test.sh",
+               },
+               Env: os.Environ(),
+       }
+
+       if hook.Path == "" {
+       }
+       spec.Hooks = &specs.Hooks{
+               Prestart: []specs.Hook{hook},
+       }
+
+       // stat, _ := os.Stat("/usr/bin/nvidia-container-runtime-hook")
+       // panic(fmt.Sprintf("stat in setupBundle: %+v\n", stat))
+
        // write the updated config
        configBytes, err = json.Marshal(spec)
        if err != nil {
@@ -460,6 +539,9 @@ func setupBundle() int {
                return 1
        }
```
- Seems that we need some implementation to cover the host device logic first, so I'm going around that for now.


Testing scenario is as follows:
- GPU-enabled instance in Paperspace (Quadro M4000).
- Built Docker from source and ran the following command multiple lines while debugging logic:
`docker run -it --gpus all nvidia/cuda:11.4.0-base-ubuntu20.04 nvidia-smi`.
- On the Dagger side, created a test that look as follows (`core/integration/gpu_test.go`) to inspect the flow:
```go
package core

import (
	"context"
	"fmt"
	"testing"

	"dagger.io/dagger"
	"github.com/stretchr/testify/require"
)

func TestGPUExperiment(t *testing.T) {
	ctx := context.Background()
	c, err := dagger.Connect(ctx)
	require.NoError(t, err)
	defer c.Close()

	ctr := c.Container().
		From("nvidia/cuda:11.4.0-base-ubuntu20.04")

	fmt.Printf("ctr: %+v\n", ctr)

	contents, err := ctr.
		WithExec([]string{"nvidia-smi"}).
		Stdout(ctx)

	require.NoError(t, err)
	fmt.Println(contents)
}
```

**Comment by matiasinsaurralde** [comment]
I've been experimenting with `runc` and wrapped up this repo with the minimal requirements to get this working: https://github.com/matiasinsaurralde/runc-nvidia-test

However I'm not yet able to make Dagger resemble the above behavior. [Here's some code](https://github.com/dagger/dagger/compare/main...matiasinsaurralde:dagger:gpu-exploration?expand=1) and notes:
- Not optimal but useful for testing: first mage copies Nvidia dependencies from the host to the dev engine container.
- Then we attempt to override the `config.json` with a minimal one like in `runc-nvidia-test`. `rootfs` path is taken from the original `config.json`.

To test:
- Run `./hack/dev bash`.
- Run `go test -v -count 1 -timeout 30s -run=TestGPUExperiment ./core/integration`:
```sh
paperspace@psal6i8au:~/go/src/github.com/dagger/dagger$ go test -v -count 1 -timeout 1000s -run=TestGPUExperiment ./core/integration
=== RUN   TestGPUExperiment
ctr: &{q:0xc00042e580 c:{Client:0xc0004b0ff0} endpoint:<nil> envVariable:<nil> exitCode:<nil> export:<nil> hostname:<nil> id:<nil> imageRef:<nil> label:<nil> platform:<nil> publish:<nil> stderr:<nil> stdout:<nil> sync:<nil> user:<nil> workdir:<nil>}
    gpu_test.go:27: 
        	Error Trace:	/home/paperspace/go/src/github.com/dagger/dagger/core/integration/gpu_test.go:27
        	Error:      	Received unexpected error:
        	            	input:1: container.from.withExec.stdout process "ls /" did not complete successfully: exit code: 1
        	            	
        	            	Stdout:
        	            	
        	            	Stderr:
        	            	runc run failed: unable to start container process: error during container init: error running hook #0: fork/exec /usr/bin/nvidia-container-runtime-hook: no such file or directory
        	            	Please visit https://dagger.io/help#go for troubleshooting guidance.
        	Test:       	TestGPUExperiment
--- FAIL: TestGPUExperiment (4.39s)
FAIL
FAIL	github.com/dagger/dagger/core/integration	4.411s
FAIL

```
- Guessing this behavior is caused by differences between `runc` container creation happening on dev engine vs container creation on the host. Still debugging possible causes.

**Comment by matiasinsaurralde** [comment]
I've found that using Alpine as the base image ([Alpine uses `musl`](https://www.linkedin.com/pulse/musl-libc-alpines-greatest-weakness-rogan-lynch/) and Nvidia tooling depends on `glibc`) for the engine causes a lot of issues. There's a relevant ticket here: https://github.com/NVIDIA/nvidia-container-runtime/issues/53

Another related ticket: https://github.com/NVIDIA/nvidia-container-toolkit/issues/71

Besides what's mentioned in the ticket I've found that [`gcompat`](https://pkgs.alpinelinux.org/package/edge/community/x86_64/gcompat) and a combination of other Alpine compatibility packages result in a workaround and have managed to run some of the Nvidia tooling with Alpine:
```sh
$ apk add gcompat libc6-compat
$ export LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu
$ nvidia-smi # works fine
```

Some interesting points:
- Should Dagger consider switching to Ubuntu base images for compatibility for scenarios like this? If workarounds like the above work fine we could still keep Alpine around and ship those compat packages by default.
- At the same time this might be a very specific scenario. The Nvidia hook is invoked by `runc` before the container is started so it runs on the host -Alpine- and not in the newly created container that has an Ubuntu base image `nvidia/cuda:11.4.0-base-ubuntu20.04` which is what I've been testing on. This scenario is probably not a common one.
- I'm still going through shim adjustments to see if we can run this tooling inside the Dagger pipeline. It was useful to explore the scenario in this way so that we can get a notion of what this hook does, what files are required, etc.

**Comment by matiasinsaurralde** [comment]
I was finally able to get the hook running properly, the code is in [this branch](https://github.com/dagger/dagger/compare/main...matiasinsaurralde:dagger:gpu-access-2?expand=1). I've added a quick integration test that queries GPUs in both the host and the Dagger container, then ensures that we're getting the same output:
```go
package core

import (
	"context"
	"os/exec"
	"testing"

	"dagger.io/dagger"
	"github.com/stretchr/testify/require"
)

func TestGPUAccess(t *testing.T) {
	ctx := context.Background()
	c, err := dagger.Connect(ctx)
	require.NoError(t, err)
	defer c.Close()

	// Query nvidia-smi on the host:
	hostCmd := exec.Command("nvidia-smi", "-L")
	hostOutput, err := hostCmd.Output()
	require.NoError(t, err)
	require.NotEmpty(t, hostOutput)

	// Query the same on the Dagger container and compare output:
	ctr := c.Container().From("nvidia/cuda:11.7.0-base-ubuntu20.04")
	contents, err := ctr.
		WithExec([]string{"nvidia-smi", "-L"}).
		Stdout(ctx)
	require.NoError(t, err)
	require.Equal(t, string(hostOutput), contents)
}
```

<img width="1256" alt="smi_dagger" src="https://github.com/dagger/dagger/assets/20110/cf9ad917-1ea4-48e9-ab93-51f1f47333c9">

**Summary:**
- Replaced dev engine with Ubuntu so that we can overcome `glibc` vs `musl` issues related to Alpine.
- Added some dirty mounts at engine initialization and shim level to inject host dependencies like shared libraries, etc. I'm wandering if it's possible to improve this part so will be researching more about `runc`.
- Some `nvidia-container-cli` subcommands return capability errors so I will look into [my standalone `runc` spec](https://github.com/matiasinsaurralde/runc-nvidia-test/blob/main/config.json) to see if there's something that can be taken from it. Cap errors are reported here too: https://github.com/NVIDIA/libnvidia-container/issues/49
- We should be able to pass `NVIDIA_VISIBLE_DEVICES` at spec level in order to trigger a behavior similar to the `--gpus` flag.

**Comment by sipsma** [comment]
This is great @matiasinsaurralde! Very exciting to see this working.

> Replaced dev engine base image with Ubuntu so that we can overcome glibc vs musl issues related to Alpine.

This SGTM, but it would also be nice to do a quick experiment with using `cgr.dev/chainguard/wolfi-base:latest`, which has `apk` as we were previously using but also uses glibc and is smaller than the `ubuntu:latest` image (`wolfi-base` is 10MB unpacked, `ubuntu` is 69MB).

If it doesn't work for whatever reason, no biggie, we can optimize in the future.

> Added some dirty mounts at engine initialization and shim level to inject host dependencies like shared libraries, etc. I'm wandering if it's possible to improve this part so will be researching more about runc.

Do you think for the "autoprovisioned" engine in docker (i.e. [here](https://github.com/dagger/dagger/compare/main...matiasinsaurralde:dagger:gpu-access-2?expand=1#diff-a81a610a4cd1cce31473f5175bc1df43001f3a1d85d010577195aa52d5025ac8R323)), we could instead just pass the `--gpus all` flag to `docker run`? I thought that would create the same result as what you did, but could be missing something.

**Comment by samalba** [comment]
> * We should be able to pass `NVIDIA_VISIBLE_DEVICES` at spec level in order to trigger a behavior similar to the `--gpus` flag.

Is this a different behavior than the docker `--gpu` arg? I would assume that it's a better UX to disable the driver capability by default and enable it only if a call like `.WithGPU(<arg>)` is passed. At the very least, it's important to be able to control the GPU visibility.

**Comment by matiasinsaurralde** [comment]
> > * We should be able to pass `NVIDIA_VISIBLE_DEVICES` at spec level in order to trigger a behavior similar to the `--gpus` flag.
> 
> Is this a different behavior than the docker `--gpu` arg? I would assume that it's a better UX to disable the driver capability by default and enable it only if a call like `.WithGPU(<arg>)` is passed. At the very least, it's important to be able to control the GPU visibility.

@samalba Same behavior, basically Docker sets the value of that environment variable based on GPUs that were specified with the CLI flag: https://github.com/moby/moby/blob/master/daemon/nvidia_linux.go#L54

**Comment by mircubed** [comment]
Thanks for the update at the community call @matiasinsaurralde! For anyone that missed it, you can check out the recording here and add any additional questions that you might have in the demo thread - https://discord.com/channels/707636530424053791/1124096404063789229/1124096404063789229 

(you will need to join the Dagger Discord to view the thread)

**Comment by matiasinsaurralde** [comment]
I've spent the past weeks polishing the code and digging deeper into multi GPU scenarios (e7e16f2697d9590d2e191af8073fc0f37b6b05f9). The latest commits in this branch include:
- `WithGPU` support with the underlying logic to pass the appropriate parameters to shim.
- shim changes to inject the environment variables expected by Nvidia.
- SDK changes to support `WithGPU`.
- Bumped the CUDA image that we were previously using as support for it was dropped -image was removed from DockerHub-.
- Rebased everything and fixed conflicts to keep up with latest `main`.
- Extended GPU tests to cover the above functionality:
```go
package core

import (
	"context"
	"os/exec"
	"regexp"
	"strings"
	"testing"

	"dagger.io/dagger"
	"github.com/stretchr/testify/require"
)

const (
	cudaImage = "nvidia/cuda:11.7.1-base-ubuntu20.04"
)

var (
	// uuidRegex is used to match GPU UUID in nvidia-smi output:
	uuidRegex = regexp.MustCompile(`GPU-\w{8}-\w{4}-\w{4}-\w{4}-\w{12}`)
)

func TestGPUAccess(t *testing.T) {
	ctx := context.Background()
	c, err := dagger.Connect(ctx)
	require.NoError(t, err)
	defer c.Close()

	// Query nvidia-smi on the host:
	hostNvidiaCmd := exec.Command("nvidia-smi", "-L")
	hostNvidiaOutput, err := hostNvidiaCmd.Output()
	hostNvidiaOutputStr := string(hostNvidiaOutput)
	require.NoError(t, err)
	require.NotEmpty(t, hostNvidiaOutput)

	// Query the same on the Dagger container and compare output:
	ctr := c.Container().From(cudaImage)
	contents, err := ctr.
		// WithGPU(dagger.ContainerWithGPUOpts{Devices: "GPU-5d8950fe-17a6-2fa7-9baa-afa83bba0e2b"}).
		WithGPU(dagger.ContainerWithGPUOpts{Devices: "all"}).
		WithExec([]string{"nvidia-smi", "-L"}).
		Stdout(ctx)
	require.NoError(t, err)
	require.Equal(t, hostNvidiaOutputStr, contents)

	t.Run("use specific GPU", func(t *testing.T) {
		var gpus []string

		// Take host output and get GPU IDs:
		for _, ln := range strings.Split(hostNvidiaOutputStr, "\n") {
			matches := uuidRegex.FindAllString(ln, 1)
			if len(matches) == 0 {
				continue
			}
			gpus = append(gpus, matches[0])
		}

		if len(gpus) <= 1 {
			t.Fatal("this test requires at least 2 GPUs to run")
		}

		// Pick first GPU and initialize a Dagger container for it:
		ctr := c.Container().From(cudaImage)
		contents, err := ctr.
			WithGPU(dagger.ContainerWithGPUOpts{Devices: gpus[0]}).
			WithExec([]string{"nvidia-smi", "-L"}).
			Stdout(ctx)
		require.NoError(t, err)

		// Assert that only the first GPU is present in the output
		// and ensure none of the other GPUs are:
		require.Contains(t, contents, gpus[0])
		for _, gpu := range gpus[1:] {
			require.NotContains(t, contents, gpu)
		}
	})
}
```

I did some experimentation around these topics too:
- Autoprovisioning the Dagger engine initial run with GPU flags: I think this didn't ressemble the behavior we're trying to achieve here (enabling access to specific GPUs on the Dagger engine container isn't the same as enabling access to GPUs i the `runc` spawned containers created during pipeline execution). Happy to discuss further now that I know more about the internals.
- `wolfi-base` didn't work out of the box, need to dig deeper. @sipsma 

Missing items to solve:
- Can we avoid mounting all Nvidia dependencies -or at least simplify it and mount less stuff- from the host into the container? What about CUDA versions? E.g. host is running Cuda 11.6, container is running Cuda 11.7. May sound confusing but it's related to the fact that the hook runs in a specific environment before the container starts.
- Current `hack/dev` is running with `export _EXPERIMENTAL_DAGGER_SERVICES_DNS=0` as this feature seems broken with Ubuntu, we need to either fix the issue or continue testing `wolfi-base` to see if it's a better fit.

**Comment by mircubed** [comment]
For those of you following this issue, @matiasinsaurralde will join us again on **September 7th during the Dagger Community call to share the latest updates on GPU access.** 

You can register here: https://dagger-io.zoom.us/webinar/register/4416686668409/WN_USQjVBGXT0SWhNMvqYVvCA#/registration

**Comment by mircubed** [comment]
If anyone would like an update on this issue, I suggest watching @matiasinsaurralde latest update from the community call - https://youtu.be/n9fBSChDvVY

**Comment by matiasinsaurralde** [comment]
For anyone following this issue, here's the latest (service containers usage fixed, improved SDK calls, added some Pytoch tests): https://github.com/dagger/dagger/pull/5605#issuecomment-1726957488

**Comment by mircubed** [comment]
@samalba @matiasinsaurralde should we close this issue now that the PR is merged?

**Comment by gerhard** [comment]
Yes! Thank you for mentioning @mircubed, closing now.
